google-cloud-pubsub/lib/google/cloud/pubsub/publisher.rb

```diff
@@ -0,0 +1,205 @@
+          @service = service
+          @max_bytes = max_bytes
+          @max_messages = max_messages
+          @timer_delay = timer_delay
+
+          @queue = {}
+          @thread_pool = Concurrent::FixedThreadPool.new(
+            [2, Concurrent.processor_count * 2].max,
+            fallback_policy: :caller_runs
+          )
+          @timer_task = Concurrent::TimerTask.new(
+            execution_interval: @timer_delay) { flush }
+
+          # init MonitorMixin
+          super()
+        end
+
+        def publish topic_name, data = nil, attributes = {}, &block
+          topic_name = service.topic_path topic_name
+          msg = create_pubsub_message data, attributes
+
+          synchronize do
+            batch = @queue[topic_name]
+            if batch.nil?
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            elsif !batch.try_add(msg, block, max_messages: max_messages,
+                                             max_bytes: max_bytes)
+              # We hit a threshold, publish existing batch.
+              publish_batch_async topic_name, batch
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            end
+            @queue[topic_name] = batch
+          end
+          nil
+        end
+
+        def start
+          @timer_task.execute
+        end
+
+        def stop
+          flush
+          @timer_task.shutdown
+        end
+
+        def started?
+          @timer_task.running?
+        end
+
+        def stopping?
+          @timer_task.shuttingdown?
+        end
+
+        def stopped?
+          @timer_task.shutdown?
+        end
+
+        def wait timeout = nil
+          @timer_task.wait_for_termination timeout
+        end
+
+        def flush
+          synchronize do
+            @queue.each do |topic_name, batch|
+              publish_batch_async topic_name, batch
+            end
+            @queue = {}
+          end
+        end
+
+        protected
+
+        def create_pubsub_message data, attributes
+          attributes ||= {}
+          if data.is_a?(::Hash) && attributes.empty?
+            attributes = data
+            data = nil
+          end
+
+          # Convert IO-ish objects to strings
+          if data.respond_to?(:read) && data.respond_to?(:rewind)
+            data.rewind
+            data = data.read
+          end
+          # Convert data to encoded byte array to match the protobuf defn
+          data = String(data).force_encoding("ASCII-8BIT")
+
+          # Convert attributes to strings to match the protobuf definition
+          attributes = Hash[attributes.map { |k, v| [String(k), String(v)] }]
+
+          Google::Pubsub::V1::PubsubMessage.new data: data,
+                                                attributes: attributes
+        end
+
+        def publish_batch topic_name, batch
+          grpc = service.publish_messages topic_name, batch.messages
```

Comment:
The retry logic is part of the GAPIC layer, which is called within `service.publish_messages`.